Add probes for memfd_create and friends

stanek-michal opened the pull request:
New events added:

EBPF_EVENT_FILE_MEMFD_OPEN
EBPF_EVENT_FILE_SHMEM_OPEN
EBPF_EVENT_PROCESS_MEMFD_CREATE
EBPF_EVENT_PROCESS_SHMGET
EBPF_EVENT_PROCESS_PTRACE
EBPF_EVENT_PROCESS_LOAD_MODULE

Also added new fields to process exec event

nicholasberlin commented on GPL/Events/File/Probe.bpf.c:
```diff
@@ -216,6 +216,60 @@ int BPF_KPROBE(kprobe__vfs_unlink)
+
+    // skip event if prefix is specified and file path does not start with it
+    if (path_prefix) {
+        if ((path_prefix_len > 0) && (size >= path_prefix_len)) {
```
Why check `size >= path_prefix_len`? I think the current value in `size` is the cgroup path length.

stanek-michal replied:
thanks for spotting this, it's a rebase artifact, 'path' was supposed to be right above the path_prefix check

nicholasberlin commented on GPL/Events/File/Probe.bpf.c:
```diff
@@ -216,6 +216,60 @@ int BPF_KPROBE(kprobe__vfs_unlink)
+    // skip event if prefix is specified and file path does not start with it
+    if (path_prefix) {
+        if ((path_prefix_len > 0) && (size >= path_prefix_len)) {
+            if (is_equal_prefix(field->data, path_prefix, path_prefix_len))
```
Not sure about `field->data` here.

nicholasberlin commented on GPL/Events/Process/Probe.bpf.c:
```diff
@@ -95,6 +98,30 @@ int BPF_PROG(sched_process_exec,
     ebpf_cred_info__fill(&event->creds, task);
     ebpf_ctty__fill(&event->ctty, task);
 
+    if (!binprm)
```
-- bikeshedding --
This if block coud be moved to the beginning of the function to avoid filling pid, cred, and ctty info.
Also, other guard logic uses `goto out`, which in this case is equivalent to `return 0`, but I think they should be consistent.

All that said, I'm guessing this will effectively always be false so there will never be any efficiency gains from moving it.
-- /bikeshedding --

nicholasberlin commented on GPL/Events/Process/Probe.bpf.c:
```diff
@@ -222,6 +249,212 @@ int tracepoint_syscalls_sys_exit_setsid(struct trace_event_raw_sys_exit *args)
+    field = ebpf_vl_field__add(&event->vl_fields, EBPF_VL_FIELD_FILENAME);
+    size  = bpf_probe_read_user_str(field->data, PATH_MAX, ex_args->uname);
+    if (size < 0)
+        return 1;
```
What does returning 1 do?

stanek-michal replied:
might have been a typo, but either way as far as I know the return value is not considered

nicholasberlin reviewed (commented):
Nice work. A few questions.

stanek-michal commented on GPL/Events/Helpers.h:
```diff
@@ -274,4 +278,15 @@ static bool is_consumer()
 }
 
+// compares first 'len' characters of str1 and str2, returns 1 if equal
+static int is_equal_prefix(const char *str1, const char *str2, int len)
```
I'm not 100% sure, but even if it didn't reject a case without fixed size arrays, it would probably just return 0 on a dereference if it was a bad page.

haesbaert replied:
I feel uneasy with this, and relying on the verifier solely seems not awesome, here is a valid alternative:
```
static int
strncmp(const char *s1, const char *s2, size_t n)
{

	if (n == 0)
		return (0);
	do {
		if (*s1 != *s2++)
			return (*(unsigned char *)s1 - *(unsigned char *)--s2);
		if (*s1++ == 0)
			break;
	} while (--n != 0);
	return (0);
}

// compares first 'len' characters of str1 and str2, returns 1 if equal
// NOTE: no bounds check, assumes use under eBPF verifier
static int is_equal_prefix(const char *str1, const char *str2, int len)
{
	return strncmp(str1, str2, len);
}

```

Compiles and runs with `EventsTrace --file-memfd-open --file-shmem-open --file-modify`

stanek-michal replied:
lgtm, I'll replace it

stanek-michal replied:
added at the top

stanek-michal commented on GPL/Events/Process/Probe.bpf.c:
```diff
@@ -222,6 +249,212 @@ int tracepoint_syscalls_sys_exit_setsid(struct trace_event_raw_sys_exit *args)
+    // memfd filename
+    field = ebpf_vl_field__add(&event->vl_fields, EBPF_VL_FIELD_FILENAME);
+    size  = bpf_probe_read_user_str(field->data, PATH_MAX, ex_args->uname);
+    if (size < 0)
```
better change to <= 0

nicholasberlin reviewed (approved):
LGTM if it works

mjwolf reviewed (approved):
The multi-kernel tester has a failing test with this, I didn't look into the exact cause, but this LGTM, once the tests are fixed

stanek-michal commented:
> The multi-kernel tester has a failing test with this, I didn't look into the exact cause, but this LGTM, once the tests are fixed

turned out it was a missing `break` in a switch block in EventsTrace that sometimes would trigger memory corruption and fail the multi-kernel tests. Fixed now and rebased, rerunning CI

haesbaert commented on GPL/Events/EbpfEventProto.h:
```diff
@@ -216,6 +232,10 @@ struct ebpf_process_exec_event {
     char comm[TASK_COMM_LEN];
+    unsigned int inode_nlink;
+    bool is_setuid;
+    bool is_setgid;
```
Can we not use bool on things that show up on the wire, if you look at the rest of the file it uses properly sized types: uint32_t, int32_t and family, it even avoids using a single int.

haesbaert replied:
Can we condense the 3 members into a uin32_t and use a flag for each bit?

haesbaert replied:
I've made this change that I think you can merge into this PR, my github-fu is severely limited, and as I said earlier, I ran out of goats:
https://github.com/elastic/ebpf/pull/194

stanek-michal replied:
agreed overall, there is an enum somewhere technically so it's not 100% portable but considering we want to be able to interface this from different languages, it's worth paying attention to

haesbaert replied:
Agreed, it's more for harmony with the rest than being portable, but it does help with stuff like someone using 
Go and having to figure out what is the size of bool.

haesbaert commented on GPL/Events/Process/Probe.bpf.c:
```diff
@@ -97,6 +103,27 @@ int BPF_PROG(sched_process_exec,
+    struct file *f        = BPF_CORE_READ(binprm, file);
+    struct inode *f_inode = BPF_CORE_READ(f, f_inode);
+    event->is_setuid      = (BPF_CORE_READ(f_inode, i_mode) & S_ISUID) ? true : false;
+    event->is_setgid      = (BPF_CORE_READ(f_inode, i_mode) & S_ISGID) ? true : false;
```
nitpick, `&` has precedence, I'd zap the parenthesis

haesbaert replied:
you can also do:
`event->is_setuid = !!(BPF_CORE_READ(f_inode, i_mode) & S_ISUID)`

stanek-michal replied:
that's one thing I do not agree with BSD folks, I like my parentheses and hate requiring everyone to know all precedence rules

haesbaert replied:
Fine by me, please disregard the other comments regarding precedence.
Now I'm in an existential crisis, am I still BSD folk, not even god knows :rofl: .

haesbaert commented on GPL/Events/Process/Probe.bpf.c:
```diff
@@ -226,6 +253,212 @@ int tracepoint_syscalls_sys_exit_setsid(struct trace_event_raw_sys_exit *args)
+    pid_t curr_tgid = BPF_CORE_READ(task, tgid);
+
+    // ignore if process is child of init/systemd/whatever
+    if ((1 == curr_tgid) || (2 == curr_tgid) || (1 == ppid) || (2 == ppid))
```
can we avoid the 1 == curr_tgid, the rest of the repository uses curr_tgid == 1, I know this is endpoint leakage :D

haesbaert replied:
can also zap the parenthesis, as == has precedence.

haesbaert commented on GPL/Events/Process/Probe.bpf.c:
```diff
@@ -226,6 +253,212 @@ int tracepoint_syscalls_sys_exit_setsid(struct trace_event_raw_sys_exit *args)
+        long shmflg;
+    };
+    struct shmget_args *ex_args = (struct shmget_args *)ctx;
+
```
blank line

haesbaert commented on GPL/Events/Process/Probe.bpf.c:
```diff
@@ -226,6 +253,212 @@ int tracepoint_syscalls_sys_exit_setsid(struct trace_event_raw_sys_exit *args)
+/* executable */
+#define MFD_EXEC 0x0010U
+    event->flags            = ex_args->flags;
+    event->flag_cloexec     = (event->flags & MFD_CLOEXEC) ? true : false;
```
why do we have to express every flag isolated again in a member, can't the user just or the original flags?

haesbaert commented on GPL/Events/EbpfEventProto.h:
```diff
@@ -271,6 +291,42 @@ struct ebpf_process_setgid_event {
+struct ebpf_process_memfd_create_event {
+    struct ebpf_event_header hdr;
+    struct ebpf_pid_info pids;
+    unsigned int flags; // memfd_create flags
+    bool flag_cloexec;
+    bool flag_allow_seal;
+    bool flag_hugetlb;
+    bool flag_noexec_seal;
+    bool flag_exec;
+
```
As noted somewhere else, why do we express flags again? Those 5 bits are in flags already, so why do we need to place them alone again?
I'm just eyeballing, so apologies if I missed something.

haesbaert reviewed (changes requested):
.

haesbaert commented on GPL/Events/Process/Probe.bpf.c:
```diff
@@ -226,6 +257,195 @@ int tracepoint_syscalls_sys_exit_setsid(struct trace_event_raw_sys_exit *args)
+    if (ebpf_events_is_trusted_pid())
+        goto out;
+
+    struct ebpf_process_load_module_event *event = get_event_buffer();
+    if (!event)
+        goto out;
+
+    event->hdr.type = EBPF_EVENT_PROCESS_LOAD_MODULE;
+    event->hdr.ts   = bpf_ktime_get_ns();
+
+    struct task_struct *task = (struct task_struct *)bpf_get_current_task();
+
+    if (is_kernel_thread(task))
+        goto out;
```
```suggestion
    struct task_struct *task = (struct task_struct *)bpf_get_current_task();

    if (is_kernel_thread(task))
        goto out;

    struct ebpf_process_load_module_event *event = get_event_buffer();
    if (!event)
        goto out;

    event->hdr.type = EBPF_EVENT_PROCESS_LOAD_MODULE;
    event->hdr.ts   = bpf_ktime_get_ns();

```

Rest of the source checks task before getting event, I think we should keep the same pattern, pattern are goods and make for a healthy breakfast.

stanek-michal replied:
done

haesbaert commented on GPL/Events/Process/Probe.bpf.c:
```diff
@@ -226,6 +257,195 @@ int tracepoint_syscalls_sys_exit_setsid(struct trace_event_raw_sys_exit *args)
+    if (ebpf_events_is_trusted_pid())
+        goto out;
+
+    struct task_struct *task = (struct task_struct *)bpf_get_current_task();
+    pid_t curr_tgid          = BPF_CORE_READ(task, tgid);
+    pid_t child_ppid         = BPF_CORE_READ(child, group_leader, real_parent, tgid);
+    pid_t child_tgid         = BPF_CORE_READ(child, tgid);
+
+    if (is_kernel_thread(task))
+        goto out;
+
```
```suggestion
    struct task_struct *task = (struct task_struct *)bpf_get_current_task();
    if (is_kernel_thread(task))
        goto out;

    pid_t curr_tgid          = BPF_CORE_READ(task, tgid);
    pid_t child_ppid         = BPF_CORE_READ(child, group_leader, real_parent, tgid);
    pid_t child_tgid         = BPF_CORE_READ(child, tgid);
```
Don't bother reading things if we're about to bail

stanek-michal replied:
done

haesbaert commented on cmake/modules/setup.cmake:
```diff
@@ -65,6 +65,8 @@ function(ebpf_static_binary target)
 
     add_executable(${target} ${EBPF_BIN_SOURCES})
 
+    target_compile_options(${target} PUBLIC -g -Wall -Wextra)
```
does this apply to the whole repository?
I'm not very familiar with cmake, but I've added a bunch of flags in `non-GPL/CMakeLists.txt`, do they get merged?

stanek-michal replied:
I think this only applies to calls to ebpf_static_binary, like this one:
`ebpf_static_binary(EventsTrace`
so EventsTrace will have debug symbols.

As for your change, not sure but best way is to look at the compilation output and find the flags

haesbaert reviewed (changes requested):
Nice work, some very minor thingies

stanek-michal commented on GPL/Events/EbpfEventProto.h:
```diff
@@ -271,6 +286,56 @@ struct ebpf_process_setgid_event {
+struct ebpf_process_ptrace_event {
+    struct ebpf_event_header hdr;
+    struct ebpf_pid_info pids;
+    int32_t child_pid;
```
done

stanek-michal commented on GPL/Events/EbpfEventProto.h:
```diff
@@ -271,6 +286,56 @@ struct ebpf_process_setgid_event {
+struct ebpf_process_shmget_event {
+    struct ebpf_event_header hdr;
+    struct ebpf_pid_info pids;
+    int32_t key;
```
done

stanek-michal commented on GPL/Events/Process/Probe.bpf.c:
```diff
@@ -97,6 +103,31 @@ int BPF_PROG(sched_process_exec,
+    struct path p              = BPF_CORE_READ(binprm, file, f_path);
+    struct dentry *curr_dentry = BPF_CORE_READ(&p, dentry);
+    struct qstr component      = BPF_CORE_READ(curr_dentry, d_name);
+    char buf_filename[8]       = {0};
```
done

haesbaert commented:
Looking swell!
